File: python/origami_lab/src/origami_lab/kinematics.py
"""Forward kinematics for a 1D origami strip.

Cell 0's local frame coincides with the world frame: origin at the
leading-edge midpoint, +X pointing along the chain, +Y along the hinge axis.

Cell i's frame is obtained from cell (i-1)'s frame by:
  1. translating by L_{i-1} along the cell's +X axis (to the shared edge),
  2. rotating by theta_{i-1} about the cell's +Y axis (the hinge).

This is the standard Denavit-Hartenberg pattern restricted to 1-DoF
revolute joints in a planar chain.

Numerically mirrors TypeScript ``@kinemind/core-math`` kinematics.ts.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass

import numpy as np
from numpy.typing import NDArray

from origami_lab.se3 import (
    SE3,
    se3_apply,
    se3_compose,
    se3_identity,
    rot,
    trans,
    quat_to_mat3,
    Vec3,
    Quat,
)
from origami_lab.strip import CELL_Y_AXIS, StripConfig, StripState

logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class CellPose:
    """World-frame pose of a single origami cell.

    Attributes:
        frame: SE3 pose of the cell's local frame relative to world.
        position: Leading-edge midpoint position shape (3,).
    """

    frame: SE3
    position: NDArray[np.float64]


@dataclass(frozen=True)
class KinematicsResult:
    """Output of forward_kinematics.

    Attributes:
        cells: Ordered list of CellPose for each cell.
        centroids: World-frame centroid of each cell (leading + trailing edges averaged).
        positions: (N,3) array of cell leading-edge midpoints.
        quats: (N,4) array of cell orientation quaternions (w,x,y,z).
        rotations: (N,3,3) array of cell rotation matrices.
    """

    cells: tuple[CellPose, ...]
    centroids: tuple[NDArray[np.float64], ...]
    positions: NDArray[np.float64]   # shape (N, 3)
    quats: NDArray[np.float64]       # shape (N, 4)
    rotations: NDArray[np.float64]   # shape (N, 3, 3)


def forward_kinematics(
    config: StripConfig,
    state: StripState,
) -> tuple[NDArray[np.float64], NDArray[np.float64]]:
    """Compute world-frame poses for every cell. O(nCells) time.

    Args:
        config: Strip geometry (cell lengths, angle limits).
        state: Hinge angle vector with len(thetas) == config.n_hinges.

    Returns:
        Tuple (positions, quats):
            - positions: ndarray shape (N, 3), leading-edge midpoints.
            - quats: ndarray shape (N, 4), cell orientations (w,x,y,z).

    Raises:
        ValueError: if state.thetas length does not match config.n_hinges.
    """
    n = config.n_cells
    if len(state.thetas) != config.n_hinges:
        raise ValueError(
            f"forward_kinematics: expected {config.n_hinges} hinge angles, "
            f"got {len(state.thetas)}"
        )

    positions = np.empty((n, 3), dtype=np.float64)
    quats = np.empty((n, 4), dtype=np.float64)

    frame = se3_identity()

    for i in range(n):
        # position = leading-edge midpoint = se3_apply(frame, [0,0,0])
        positions[i] = se3_apply(frame, np.zeros(3))
        quats[i] = frame.q

        if i < n - 1:
            L_i = float(config.cell_lengths[i])
            theta = float(state.thetas[i])
            # Translate to shared edge, then rotate about hinge axis (+Y).
            frame = se3_compose(
                frame,
                se3_compose(
                    trans(np.array([L_i, 0.0, 0.0])),
                    rot(CELL_Y_AXIS, theta),
                ),
            )

    return positions, quats


def forward_kinematics_full(
    config: StripConfig,
    state: StripState,
) -> KinematicsResult:
    """Compute full kinematics result including centroids and rotation matrices.

    Args:
        config: Strip geometry.
        state: Hinge angle vector.

    Returns:
        KinematicsResult with cells, centroids, positions, quats, rotations.

    Raises:
        ValueError: if state shape disagrees with config.
    """
    n = config.n_cells
    if len(state.thetas) != config.n_hinges:
        raise ValueError(
            f"forward_kinematics_full: expected {config.n_hinges} hinge angles, "
            f"got {len(state.thetas)}"
        )

    cells_list: list[CellPose] = []
    centroids_list: list[NDArray[np.float64]] = []

    frame = se3_identity()

    for i in range(n):
        pos = se3_apply(frame, np.zeros(3))
        cells_list.append(CellPose(frame=frame, position=pos.copy()))
        L_i = float(config.cell_lengths[i])
        centroids_list.append(se3_apply(frame, np.array([L_i * 0.5, 0.0, 0.0])))

        if i < n - 1:
            theta = float(state.thetas[i])
            frame = se3_compose(
                frame,
                se3_compose(
                    trans(np.array([L_i, 0.0, 0.0])),
                    rot(CELL_Y_AXIS, theta),
                ),
            )

    positions = np.stack([c.position for c in cells_list])
    quats = np.stack([c.frame.q for c in cells_list])
    rotations = np.stack([quat_to_mat3(c.frame.q) for c in cells_list])

    return KinematicsResult(
        cells=tuple(cells_list),
        centroids=tuple(centroids_list),
        positions=positions,
        quats=quats,
        rotations=rotations,
    )


def cell_corners_local(cell_length: float) -> NDArray[np.float64]:
    """Return the 4 corners of a cell in its local frame (CCW from +Z).

    Mirrors TypeScript ``cellCornersLocal`` in kinematics.ts.

    Args:
        cell_length: Cell length along +X.

    Returns:
        Corners array shape (4, 3) in CCW order viewed from +Z.
    """
    half = 0.5
    return np.array(
        [
            [0.0, -half, 0.0],
            [cell_length, -half, 0.0],
            [cell_length, half, 0.0],
            [0.0, half, 0.0],
        ],
        dtype=np.float64,
    )


def cell_corners_world(
    config: StripConfig,
    result: "KinematicsResult",
) -> list[NDArray[np.float64]]:
    """Return world-frame corners of every cell.

    Mirrors TypeScript ``cellCornersWorld`` in kinematics.ts.

    Args:
        config: Strip geometry.
        result: KinematicsResult from :func:`forward_kinematics_full`.

    Returns:
        List of length n_cells; each element is an ndarray shape (4, 3).
    """
    out: list[NDArray[np.float64]] = []
    for i, cell_pose in enumerate(result.cells):
        L = float(config.cell_lengths[i])
        local = cell_corners_local(L)
        # Apply SE3 transform to each corner.
        world = np.stack([se3_apply(cell_pose.frame, local[k]) for k in range(4)])
        out.append(world)
    return out

File: packages/core-math/src/kinematics.ts
/**
 * Forward kinematics for a 1D origami strip.
 *
 * Cell 0's local frame coincides with the world frame: origin at the
 * leading-edge midpoint, +X pointing along the chain, +Y along the hinge axis.
 * Cell i's frame is obtained from cell (i-1)'s frame by:
 *   1. translating by L_{i-1} along the cell's +X axis (to the shared edge),
 *   2. rotating by theta_{i-1} about the cell's +Y axis (the hinge),
 *
 * which is the standard Denavit-Hartenberg pattern restricted to 1-DoF
 * revolute joints in a planar chain. The resulting cell positions are the
 * leading-edge midpoints of each cell.
 */

import {
  type SE3,
  SE3_IDENTITY,
  type Vec3,
  rot,
  se3Apply,
  se3Compose,
  trans,
  v3Add,
  v3Scale,
} from "./se3.js";
import {
  CELL_X_AXIS,
  CELL_Y_AXIS,
  CELL_Z_AXIS,
  type StripConfig,
  type StripState,
} from "./strip.js";

export interface CellPose {
  /** Pose of the cell's local frame relative to world. */
  readonly frame: SE3;
  /** Leading-edge midpoint position. */
  readonly position: Vec3;
}

export interface KinematicsResult {
  readonly cells: readonly CellPose[];
  /** Centroids of every cell (leading + trailing edges averaged). */
  readonly centroids: readonly Vec3[];
}

/**
 * Compute world-frame poses for every cell. Runs in O(nCells) time.
 * Throws if the state shape disagrees with the strip config.
 */
export function forwardKinematics(
  config: StripConfig,
  state: StripState,
): KinematicsResult {
  if (state.thetas.length !== config.nCells - 1) {
    throw new Error(
      `forwardKinematics: expected ${config.nCells - 1} hinge angles, got ${state.thetas.length}`,
    );
  }

  const cells: CellPose[] = [];
  const centroids: Vec3[] = [];
  let frame: SE3 = SE3_IDENTITY;

  for (let i = 0; i < config.nCells; i++) {
    const position = se3Apply(frame, [0, 0, 0]);
    cells.push({ frame, position });
    const cellLength = config.cellLengths[i] ?? 0;
    centroids.push(se3Apply(frame, [cellLength * 0.5, 0, 0]));

    if (i < config.nCells - 1) {
      const Li = config.cellLengths[i] ?? 0;
      const theta = state.thetas[i] ?? 0;
      // Translate to the shared edge, then rotate about the hinge axis.
      frame = se3Compose(
        frame,
        se3Compose(trans([Li, 0, 0]), rot(CELL_Y_AXIS, theta)),
      );
    }
  }

  return { cells, centroids };
}

/**
 * Local-frame corner offsets of a unit-width cell with edge L along +X.
 * Returns the four corners in CCW order viewed from +Z.
 */
export function cellCornersLocal(cellLength: number): readonly Vec3[] {
  const half = 0.5;
  return [
    [0, -half, 0],
    [cellLength, -half, 0],
    [cellLength, half, 0],
    [0, half, 0],
  ];
}

/**
 * World-frame corners of every cell, suitable for collision tests and rendering.
 */
export function cellCornersWorld(
  config: StripConfig,
  result: KinematicsResult,
): readonly (readonly Vec3[])[] {
  const out: Vec3[][] = [];
  for (let i = 0; i < config.nCells; i++) {
    const cellLength = config.cellLengths[i] ?? 0;
    const cell = result.cells[i];
    if (!cell) continue;
    const local = cellCornersLocal(cellLength);
    out.push(local.map((p) => se3Apply(cell.frame, p)));
  }
  return out;
}

/**
 * Axis-aligned bounding box from a cloud of points.
 */
export function computeAabb(points: readonly Vec3[]): {
  min: Vec3;
  max: Vec3;
} {
  if (points.length === 0) {
    return { min: [0, 0, 0], max: [0, 0, 0] };
  }
  const first = points[0]!;
  let minX = first[0];
  let minY = first[1];
  let minZ = first[2];
  let maxX = first[0];
  let maxY = first[1];
  let maxZ = first[2];
  for (let i = 1; i < points.length; i++) {
    const p = points[i]!;
    if (p[0] < minX) minX = p[0];
    if (p[1] < minY) minY = p[1];
    if (p[2] < minZ) minZ = p[2];
    if (p[0] > maxX) maxX = p[0];
    if (p[1] > maxY) maxY = p[1];
    if (p[2] > maxZ) maxZ = p[2];
  }
  return { min: [minX, minY, minZ], max: [maxX, maxY, maxZ] };
}

export function aabbOverlap(
  a: { min: Vec3; max: Vec3 },
  b: { min: Vec3; max: Vec3 },
  tol = 1e-9,
): boolean {
  return (
    a.min[0] <= b.max[0] + tol &&
    a.max[0] + tol >= b.min[0] &&
    a.min[1] <= b.max[1] + tol &&
    a.max[1] + tol >= b.min[1] &&
    a.min[2] <= b.max[2] + tol &&
    a.max[2] + tol >= b.min[2]
  );
}

/**
 * Detect self-intersection between non-adjacent cells using a broad-phase
 * AABB pass. The exact narrow-phase SAT check is intentionally omitted in
 * the v0.1 release: the AABB pass already catches every case that matters
 * for an interactive viewer, and full SAT belongs in MATH.md tests where
 * we can assert the algebraic guarantees properly.
 *
 * Returns the first colliding pair (i, j) with j > i + 1, or null.
 */
export function detectSelfIntersectionAabb(
  config: StripConfig,
  result: KinematicsResult,
): [number, number] | null {
  const corners = cellCornersWorld(config, result);
  const aabbs = corners.map(computeAabb);
  for (let i = 0; i < aabbs.length; i++) {
    for (let j = i + 2; j < aabbs.length; j++) {
      const a = aabbs[i]!;
      const b = aabbs[j]!;
      if (aabbOverlap(a, b)) {
        return [i, j];
      }
    }
  }
  return null;
}

/**
 * Numerical Jacobian of the i-th cell centroid with respect to the hinge
 * angle vector, computed by central finite differences. O(nCells^2).
 */
export function centroidJacobian(
  config: StripConfig,
  state: StripState,
  cellIndex: number,
  h = 1e-5,
): readonly Vec3[] {
  const n = state.thetas.length;
  const J: Vec3[] = [];
  for (let k = 0; k < n; k++) {
    const tPlus = [...state.thetas];
    const tMinus = [...state.thetas];
    tPlus[k] = (tPlus[k] ?? 0) + h;
    tMinus[k] = (tMinus[k] ?? 0) - h;
    const plus = forwardKinematics(config, { thetas: tPlus }).centroids[cellIndex];
    const minus = forwardKinematics(config, { thetas: tMinus }).centroids[cellIndex];
    if (!plus || !minus) {
      J.push([0, 0, 0]);
      continue;
    }
    const diff = v3Scale(
      [plus[0] - minus[0], plus[1] - minus[1], plus[2] - minus[2]],
      1 / (2 * h),
    );
    J.push(diff);
  }
  return J;
}

/** Re-export common axes for callers building stimuli on top of the strip. */
export const STRIP_AXES = {
  cellX: CELL_X_AXIS,
  cellY: CELL_Y_AXIS,
  cellZ: CELL_Z_AXIS,
} as const;

/**
 * Orientation reproduction helper. Compose all hinge rotations into a single
 * transform applied to v. Useful for unit tests that round-trip a unit vector.
 */
export function chainRotateVector(
  config: StripConfig,
  state: StripState,
  v: Vec3,
): Vec3 {
  const { cells } = forwardKinematics(config, state);
  const last = cells[cells.length - 1];
  if (!last) {
    return v;
  }
  return v3Add([0, 0, 0], se3Apply({ q: last.frame.q, t: [0, 0, 0] }, v));
}
